### PR TITLE
add: withdraw interest gains unit tests, withdraw recipient. change: …

### PR DESCRIPTION
**Pull request by ordd**
…timestamp to block number.

# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

About # (link your issue here)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes.

# Checklist:
- [ ] PR title matches follow: (Feature|Bug|Chore) Task Name
- [ ] My code follows the style guidelines of this project
- [ ] I have followed all the instructions described in the initial task (check Definitions of Done)
- [ ] I have performed a self-review of my own code
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have added reference to a related issue in the repository
- [ ] I have added a detailed description of the changes proposed in the pull request. I am as descriptive as possible, assisting reviewers as much as possible.
- [ ] I have added screenshots related to my pull request ( for frontend tasks)
- [ ] I have pasted a gif showing the feature.
- [ ] @mentions of the person or team responsible for reviewing proposed changes

### CODE REVIEW

File: stakingModel/contracts/SimpleDAIStaking.sol [hunk 1]
```diff
@@ -133,32 +139,43 @@ contract SimpleDAIStaking is DSMath, Pausable, SchemeGuard {
...
-            totalStaked; //mul by 1e10 to convert cdai precision to dai precision, div to reduce precision from 1e28 of exchange rate to 1e27 that DSMath works on
+        uint256 daiWorth = currentDAIWorth();
+        if (daiWorth < totalStaked) {
+            return (0, 0, 0);
+        }
+        uint256 daiGains = daiWorth - totalStaked;
         uint256 cdaiGains = rdiv(daiGains * 1e10, er); //mul by 1e10 to equalize precision otherwise since exchangerate is very big, dividing by it would result in 0.
         uint256 precisionLossCDaiRay = cdaiGains % 1e19; //get right most bits not covered by precision of cdai which is only 8 decimals while RAY is 27
-        cdaiGains = cdaiGains.div(1e19); //lower back to 8 decimals
+        if (cdaiGains != 0) {
```
Comment (sirpy) [inline stakingModel/contracts/SimpleDAIStaking.sol]: what for? 0 divided by a number is 0

File: stakingModel/test/SimpleDAIStaking.test.js [hunk 2]
```diff
@@ -1,368 +1,503 @@
 const SimpleDAIStaking = artifacts.require("SimpleDAIStaking");
-const GoodDollar = artifacts.require("GoodDollar");
```
Comment (sirpy) [inline stakingModel/test/SimpleDAIStaking.test.js]: restore original indentation its impossible to compare the changes

**Review by sirpy — CHANGES_REQUESTED** [review]

File: stakingModel/contracts/SimpleDAIStaking.sol [hunk 3]
```diff
@@ -112,9 +118,10 @@ contract SimpleDAIStaking is DSMath, Pausable, SchemeGuard {
         uint256 daiWithdraw = staker.stakedDAI;
         staker.stakedDAI = 0; // update balance before transfer to prevent re-entry
         totalStaked -= daiWithdraw;
+        uint256 daiActual = dai.balanceOf(address(this));
```
Comment (sirpy) [inline stakingModel/contracts/SimpleDAIStaking.sol:121]:
```suggestion
        uint256 daiActual = dai.balanceOf(address(this)); //redeem of cDAI might have some precision loss
```

Comment (sirpy) [reply stakingModel/contracts/SimpleDAIStaking.sol:121]: maybe also calculate dai balance before/after and transfer the difference?

File: stakingModel/contracts/SimpleDAIStaking.sol [hunk 4]
```diff
@@ -133,32 +140,43 @@ contract SimpleDAIStaking is DSMath, Pausable, SchemeGuard {
...
         uint256 er = cDai.exchangeRateStored();
-        uint256 daiGains = rmul(cDai.balanceOf(address(this)) * 1e10, er).div(
-            10
-        ) -
-            totalStaked; //mul by 1e10 to convert cdai precision to dai precision, div to reduce precision from 1e28 of exchange rate to 1e27 that DSMath works on
+        uint256 daiWorth = currentDAIWorth();
+        if (daiWorth < totalStaked) {
+            return (0, 0, 0);
+        }
+        uint256 daiGains = daiWorth - totalStaked;
```
Comment (sirpy) [inline stakingModel/contracts/SimpleDAIStaking.sol]:
```suggestion
        uint256 daiGains = daiWorth.sub(totalStaked);
```

File: stakingModel/contracts/SimpleDAIStaking.sol [hunk 5]
```diff
@@ -133,32 +140,43 @@ contract SimpleDAIStaking is DSMath, Pausable, SchemeGuard {
...
         uint256 cdaiGains = rdiv(daiGains * 1e10, er); //mul by 1e10 to equalize precision otherwise since exchangerate is very big, dividing by it would result in 0.
         uint256 precisionLossCDaiRay = cdaiGains % 1e19; //get right most bits not covered by precision of cdai which is only 8 decimals while RAY is 27
-        cdaiGains = cdaiGains.div(1e19); //lower back to 8 decimals
+        if (cdaiGains > 0) {
```
Comment (sirpy) [inline stakingModel/contracts/SimpleDAIStaking.sol]: add comment why

File: stakingModel/contracts/SimpleDAIStaking.sol [hunk 6]
```diff
@@ -56,11 +58,15 @@ contract SimpleDAIStaking is DSMath, Pausable, SchemeGuard {
...
-        address _fundManager
+        address _fundManager,
+        uint256 _blockInterval
+
     ) public SchemeGuard(Avatar(address(0))) {
         dai = ERC20(_dai);
         cDai = cERC20(_cDai);
         uniswap = _uniswap;
+        blockInterval = _blockInterval;
+        lastUBICollection = block.number + blockInterval;
```
Comment (sirpy) [inline stakingModel/contracts/SimpleDAIStaking.sol]: this prevents first interval withdraw, probably should be just block.number
need to add a unit test

**Review by sirpy — CHANGES_REQUESTED** [review]

File: stakingModel/test/SimpleDAIStaking.e2e.test.js [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+export const DAI_ADDRESS  = "0x4F96Fe3b7A6Cf9725f59d353F723c1bDb64CA6Aa";
+export const cDAI_ADDRESS = "0xe7bc397dbd069fc7d0109c0636d06888bb50668c";
+
+contract("SimpleDAIStaking - staking with DAI mocks", ([founder, staker]) => {
+  let dai;
+  let cDAI;
+  let simpleStaking;
+
+  before(async () => {
+    dai = await DAIMock.at(DAI_ADDRESS);
```
Comment (sirpy) [inline stakingModel/test/SimpleDAIStaking.e2e.test.js:26]: hmm does that work? using our mocks with real addresses?

File: stakingModel/test/SimpleDAIStaking.e2e.test.js [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+const GoodDollar = artifacts.require("GoodDollar");
+const DAIMock = artifacts.require("DAIMock");
+const cDAIMock = artifacts.require("cDAIMock");
+
+const fse = require("fs-extra");
+
+const BN = web3.utils.BN;
+export const BLOCK_INTERVAL = 2;
+export const NULL_ADDRESS = "0x0000000000000000000000000000000000000000";
+export const DAI_ADDRESS  = "0x4F96Fe3b7A6Cf9725f59d353F723c1bDb64CA6Aa";
```
Comment (sirpy) [inline stakingModel/test/SimpleDAIStaking.e2e.test.js:13]: describe on which network

File: stakingModel/test/SimpleDAIStaking.e2e.test.js [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+export const cDAI_ADDRESS = "0xe7bc397dbd069fc7d0109c0636d06888bb50668c";
+
+contract("SimpleDAIStaking - staking with DAI mocks", ([founder, staker]) => {
```
Comment (sirpy) [inline stakingModel/test/SimpleDAIStaking.e2e.test.js:16]: add network to parameters and dont run if not on correct network (ie if network is test/develop)

**Review by sirpy — CHANGES_REQUESTED** [review]

File: stakingModel/test/SimpleDAIStaking.e2e.test.js [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+      web3.utils.toWei("500", "gwei")
+    );
+    let totalStakedAfter = await simpleStaking.totalStaked();
+    expect((totalStakedAfter - totalStakedBefore).toString()).to.be.equal(
+      web3.utils.toWei("500", "gwei")
+    );
+    const gains = await simpleStaking.currentUBIInterest();
+    const cdaiGains = gains["0"];
+    const fundBalanceBefore = await cDAI.balanceOf(founder);
+    await simpleStaking.collectUBIInterest(founder);
```
Comment (sirpy) [inline stakingModel/test/SimpleDAIStaking.e2e.test.js:59]: are we gonna have any gains? maybe mint very large amount of DAI and wait a few blocks?
is it possible to test this on a test network in a unit test? how can we actually test cDAI generating revenues

**Review by sirpy — CHANGES_REQUESTED** [review]

File: stakingModel/test/SimpleDAIStaking.e2e.test.js (hunk 7 above)
Comment (ordd) [reply stakingModel/test/SimpleDAIStaking.e2e.test.js:26]: yes, same methods

File: stakingModel/contracts/SimpleDAIStaking.sol [hunk 11]
```diff
@@ -144,10 +147,10 @@ contract SimpleDAIStaking is DSMath, Pausable, SchemeGuard {
         if (daiWorth < totalStaked) {
             return (0, 0, 0);
         }
-        uint256 daiGains = daiWorth - totalStaked;
+        uint256 daiGains = daiWorth.sub(totalStaked);
         uint256 cdaiGains = rdiv(daiGains * 1e10, er); //mul by 1e10 to equalize precision otherwise since exchangerate is very big, dividing by it would result in 0.
         uint256 precisionLossCDaiRay = cdaiGains % 1e19; //get right most bits not covered by precision of cdai which is only 8 decimals while RAY is 27
-        if (cdaiGains > 0) {
+        if (cdaiGains > 0) { // otherwise 0
```
Comment (sirpy) [inline stakingModel/contracts/SimpleDAIStaking.sol]: the comment here is that dividing 0 gives non 0 no?

**Review by sirpy — APPROVED** [review]

File: stakingModel/test/SimpleDAIStaking.e2e.test.js (hunk 9 above)
Comment (sirpy) [reply stakingModel/test/SimpleDAIStaking.e2e.test.js:16]: ?

File: stakingModel/test/SimpleDAIStaking.e2e.test.js (hunk 10 above)
Comment (sirpy) [reply stakingModel/test/SimpleDAIStaking.e2e.test.js:59]: ?

File: stakingModel/test/SimpleDAIStaking.e2e.test.js (hunk 9 above)
Comment (sirpy) [reply stakingModel/test/SimpleDAIStaking.e2e.test.js:16]: waiting

File: stakingModel/contracts/SimpleDAIStaking.sol (hunk 11 above)
Comment (sirpy) [reply stakingModel/contracts/SimpleDAIStaking.sol]: waiting

Comment (ordd) [reply stakingModel/contracts/SimpleDAIStaking.sol]: a comment has added

File: stakingModel/test/SimpleDAIStaking.e2e.test.js (hunk 9 above)
Comment (ordd) [reply stakingModel/test/SimpleDAIStaking.e2e.test.js:16]: skipping the test if the network is test or develop

File: stakingModel/test/SimpleDAIStaking.e2e.test.js [hunk 12]
```diff
@@ -18,10 +18,10 @@ contract("SimpleDAIStaking - staking with DAI mocks", ([founder, staker]) => {
   let cDAI;
   let simpleStaking;
 
-  before(async () => {
+  before(async function() {
     let network = process.env.NETWORK;
     if(network === 'test' || network === 'develop'){
```
Comment (sirpy) [inline stakingModel/test/SimpleDAIStaking.e2e.test.js]:
```suggestion
    if(network !== 'kovan'){
```

**Pull request merged**